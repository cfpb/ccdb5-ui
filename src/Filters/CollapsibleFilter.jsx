import './CollapsibleFilter.less';
import iconMap from '../iconMap'
import PropTypes from 'prop-types'
import React from 'react';

export default class CollapsibleFilter extends React.Component {
  constructor( props ) {
    super( props );
    this.state = {
      showChildren: props.showChildren
    };

    // This binding is necessary to make `this` work in the callback
    // https://facebook.github.io/react/docs/handling-events.html
    this._toggleChildDisplay = this._toggleChildDisplay.bind( this );
  }

  _toggleChildDisplay() {
    this.setState( {
      showChildren: !this.state.showChildren
    } );
  }

  componentDidUpdate( prevProps ) {
    if ( prevProps.showChildren !== this.props.showChildren ) {
      // sync local state
      this.setState( {
        showChildren: this.props.showChildren
      } )
    }
  }

  render() {
    let composeClasses = 'o-expandable'
    if ( this.props.className ) {
      composeClasses += ' ' + this.props.className;
    }

    const buttonClasses = 'a-btn a-btn__link o-expandable_cue '

    const opened =
      <button className={ buttonClasses + 'o-expandable_cue-close' }
              onClick={this._toggleChildDisplay}>
        Hide
        { iconMap.getIcon( 'minus-round' ) }
      </button>
    const closed =
      <button className={ buttonClasses + 'o-expandable_cue-open' }
              onClick={this._toggleChildDisplay}>
        Show
        { iconMap.getIcon( 'plus-round' ) }
      </button>

    return (
      <section className={composeClasses}>
        <div className="o-expandable_header o-expandable_target">
          <h4 className="o-expandable_header-left o-expandable_label">
              {this.props.title}
          </h4>
          <span className="o-expandable_header-right o-expandable_link">
            { this.state.showChildren ? opened : closed }
          </span>
        </div>
        <p>{this.props.desc}</p>
        { this.state.showChildren ? this.props.children : null }
      </section>
    );
  }
}

CollapsibleFilter.propTypes = {
  showChildren: PropTypes.bool
}

CollapsibleFilter.defaultProps = {
  showChildren: true
}
