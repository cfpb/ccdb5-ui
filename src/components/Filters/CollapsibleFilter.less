.o-expandable {
  margin-bottom: 0;
  h4 {
    max-width: ~'calc(100% - 55px)';
    box-sizing: border-box;
    padding-right: @gutter-minimum;
  }

  .o-expandable_cue {
    &.a-btn__link {
      border: 0;

      font-size: @size-v;
    }

    .cf-icon-svg {
      padding-left: @gutter-minimum;
    }
  }

  p {
    color: @gray;
    font-size: 16px;
  }
}
