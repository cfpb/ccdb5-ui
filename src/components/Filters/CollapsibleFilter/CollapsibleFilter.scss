@use 'css/base' as *;

.o-expandable {
  margin-bottom: 0;
  h4 {
    max-width: calc(100% - 55px);
    box-sizing: border-box;
    padding-right: $gutter-minimum;
  }

  p {
    color: var(--gray);
    font-size: 16px;
  }
}
