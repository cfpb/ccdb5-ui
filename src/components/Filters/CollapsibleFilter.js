import './CollapsibleFilter.less';
import iconMap from '../iconMap';
import PropTypes from 'prop-types';
import React from 'react';

export default class CollapsibleFilter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      hasChildren: props.hasChildren,
    };

    // This binding is necessary to make `this` work in the callback
    // https://facebook.github.io/react/docs/handling-events.html
    this._toggleChildDisplay = this._toggleChildDisplay.bind(this);
  }

  _toggleChildDisplay() {
    this.setState({
      hasChildren: !this.state.hasChildren,
    });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.hasChildren !== this.props.hasChildren) {
      // sync local state
      this.setState({
        hasChildren: this.props.hasChildren,
      });
    }
  }

  render() {
    let composeClasses = 'o-expandable';
    if (this.props.className) {
      composeClasses += ' ' + this.props.className;
    }

    const buttonClasses = this.state.hasChildren
      ? '__expanded'
      : '__collapsed';

    const opened = (
      <>
        <span className="o-expandable_cue o-expandable_cue-close">
          <span className="u-visually-hidden-on-mobile">Hide</span>
          {iconMap.getIcon('minus-round')}
        </span>
      </>
    );
    const closed = (
      <>
        <span className="o-expandable_cue o-expandable_cue-open">
          <span className="u-visually-hidden-on-mobile">Show</span>
          {iconMap.getIcon('plus-round')}
        </span>
      </>
    );
    return (
      <section className={composeClasses}>
        <button
          className={
            'o-expandable_header o-expandable_target o-expandable_target' +
            buttonClasses
          }
          aria-label={`Hide ${this.props.title} filter`}
          onClick={this._toggleChildDisplay}
        >
          <h3 className="h4 o-expandable_header-left o-expandable_label">
            {this.props.title}
          </h3>
          <span className="o-expandable_header-right o-expandable_link">
            {this.state.hasChildren ? opened : closed}
          </span>
        </button>
        {this.state.hasChildren ? (
          <div className="o-expandable_content">
            <p>{this.props.desc}</p>
            {this.props.children}
          </div>
        ) : null}
      </section>
    );
  }
}

CollapsibleFilter.propTypes = {
  hasChildren: PropTypes.bool,
  className: PropTypes.string,
  title: PropTypes.string,
  desc: PropTypes.string,
  children: PropTypes.node,
};

CollapsibleFilter.defaultProps = {
  hasChildren: true,
};
