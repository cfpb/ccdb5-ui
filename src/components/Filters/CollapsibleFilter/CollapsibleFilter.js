import './CollapsibleFilter.less';
import getIcon from '../../iconMap';
import PropTypes from 'prop-types';
import { useState } from 'react';

const CollapsibleFilter = ({
  hasChildren = true,
  className = '',
  title,
  desc,
  children,
}) => {
  const [isOpen, setOpen] = useState(hasChildren);

  const opened = (
    <span className="o-expandable__cue-close" role="img" aria-label="Hide">
      {getIcon('minus-round')}
    </span>
  );
  const closed = (
    <span className="o-expandable__cue-open" role="img" aria-label="Show">
      {getIcon('plus-round')}
    </span>
  );

  return (
    <section className={`o-expandable ${className}`}>
      <button
        className="o-expandable__header"
        aria-expanded={isOpen}
        aria-label={`Hide ${title} filter`}
        onClick={() => setOpen(!isOpen)}
      >
        <h3 className="o-expandable__label">{title}</h3>
        <span className="o-expandable__cues">{isOpen ? opened : closed}</span>
      </button>
      {isOpen ? (
        <div className="o-expandable__content">
          <p>{desc}</p>
          {children}
        </div>
      ) : null}
    </section>
  );
};

CollapsibleFilter.propTypes = {
  hasChildren: PropTypes.bool,
  className: PropTypes.string,
  title: PropTypes.string,
  desc: PropTypes.string,
  children: PropTypes.node,
};

export default CollapsibleFilter;
