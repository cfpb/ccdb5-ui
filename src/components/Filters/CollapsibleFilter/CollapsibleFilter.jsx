import './CollapsibleFilter.scss';
import getIcon from '../../Common/Icon/iconMap';
import PropTypes from 'prop-types';
import { useState } from 'react';

export const CollapsibleFilter = ({
  className = '',
  title,
  desc,
  children,
}) => {
  const [isOpen, setOpen] = useState(true);

  const opened = (
    <span className="o-expandable__cue-close" role="img" aria-label="Hide">
      {getIcon('minus-round')}
    </span>
  );
  const closed = (
    <span className="o-expandable__cue-open" role="img" aria-label="Show">
      {getIcon('plus-round')}
    </span>
  );
  const label = isOpen ? `Collapse ${title} filter` : `Expand ${title} filter`;
  return (
    <section className={`o-expandable ${className}`}>
      <button
        className="o-expandable__header"
        aria-expanded={isOpen}
        aria-label={label}
        onClick={() => setOpen(!isOpen)}
      >
        <h3 className="o-expandable__label">{title}</h3>
        <span className="o-expandable__cues">{isOpen ? opened : closed}</span>
      </button>
      {isOpen ? (
        <div className="o-expandable__content">
          <p>{desc}</p>
          {children}
        </div>
      ) : null}
    </section>
  );
};

CollapsibleFilter.propTypes = {
  className: PropTypes.string,
  title: PropTypes.string.isRequired,
  desc: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};
