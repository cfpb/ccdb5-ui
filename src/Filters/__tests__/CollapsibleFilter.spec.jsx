import React from 'react';
import { mount } from 'enzyme';
import CollapsibleFilter from '../CollapsibleFilter';
import renderer from 'react-test-renderer';

describe('component:CollapsibleFilter', () => {
  describe('initial state', () => {
    it('renders without crashing', () => {
      const target = renderer.create(
        <CollapsibleFilter title="foo" desc="bar" />
      );

      let tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('hides the children when Hide is clicked', () => {
    const target = mount(<CollapsibleFilter showChildren={true} />);
    const theButton = target.find('button.o-expandable_cue');

    expect(target.state('showChildren')).toEqual(true);
    theButton.simulate('click');
    expect(target.state('showChildren')).toEqual(false);
  })

  describe('componentDidUpdate', () => {
    it( 'triggers a new state update when props change', () => {
      const props = {
        showChildren: false
      }

      const target = mount( <CollapsibleFilter { ...props } /> )
      target.setProps( { showChildren: true } )
      const sv = target.state( 'showChildren' )
      expect( sv ).toEqual( true )
    } )
  })
})
