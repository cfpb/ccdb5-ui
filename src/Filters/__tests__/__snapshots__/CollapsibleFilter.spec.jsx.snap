// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:CollapsibleFilter initial state renders without crashing 1`] = `
<section
  className="o-expandable"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      foo
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <svg
          className="cf-icon-svg"
          viewBox="0 0 1000 1200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M500 105.2c-276.1 0-500 223.9-500 500s223.9 500 500 500 500-223.9 500-500-223.9-500-500-500zm263.1 550.7H236c-27.6 0-50-22.4-50-50s22.4-50 50-50h527.1c27.6 0 50 22.4 50 50s-22.4 50-50 50z"
          />
        </svg>
      </button>
    </span>
  </div>
  <p>
    bar
  </p>
</section>
`;
